src/Refinement/RefinementTransferOperators.jl

```diff
@@ -0,0 +1,166 @@
+
+"""
+  RefinementTransferOperator
+
+  Matrix-like operator providing a transfer between two `FESpaces`
+  defined in child-parent refined grids. 
+
+  The dofs are transfered by iteratively solving a projection problem 
+
+        (uh_to,vh_to)_to = (uh_from,vh_to)_Ω
+
+  where Ω is always chosen as the finest grid (i.e the RHS is
+  always integrated without loss of accuracy). 
+"""
+struct RefinementTransferOperator{T,A,B,C} <: AbstractMatrix{T}
+  from   ::A
+  to     ::B
+  caches ::C
+
+  function RefinementTransferOperator(T,from,to,caches)
+    A = typeof(from)
+    B = typeof(to)
+    C = typeof(caches)
+    new{T,A,B,C}(from,to,caches)
+  end
+end
+
+# L2 projection (default for the transfer ops)
+Π_l2(u,v,dΩ) = ∫(v⋅u)*dΩ
+
+function RefinementTransferOperator(from::FESpace,to::FESpace; Π=Π_l2, qdegree=3)
```

Comment:
I am not wrong, you have tried to unify both the prolongation and the restriction operators in a single data type. While this is interesting in order to avoid code replication, I am not sure if this is general enough in the sense that one may want to have, for example, the interpolation operator as the prolongator operator (from coarse to fine), and its adjoint as the restriction operator (from fine to coarse). 

In such a case, the action of each operator on a function would be very different depending on the operator.

Reply:
I don't see why my implementation would be a problem... With the current code (you can refer to the tests) we have to create two operators (which can have different projections for instance) for prolongation and restriction. The only thing that is unified is the code that creates them. But how the resulting operator behaves will depend on the `from` and `to` spaces. The idea is that in a future we will not care on which of the two spaces is more or less refined, since we will have coarsening and refinement in each AMR step. We just want an operator that can go from one space to the other independently of what refinement has been applied, so there is no point in creating two.